### PR TITLE
CHAOS-733: allow users to define multiple paths for disk failure disruptions

### PR DESCRIPTION
**Pull request by aymericDD**
## What does this PR do?

- [ ] Adds new functionality
- [x] Alters existing functionality
- [ ] Fixes a bug
- [ ] Improves documentation or testing

Please briefly describe your changes as well as the motivation behind them:
- Allow users to define a list of paths for disk failure disruptions in order to disrupt multiple paths.

> **Before**

<details>

``` yaml
---
apiVersion: chaos.datadoghq.com/v1beta1
kind: Disruption
metadata:
  name: disk-failure
  namespace: chaos-demo
  annotations:
    chaos.datadoghq.com/environment: "lima"
spec:
  level: pod
  selector:
    app: demo-curl
  count: 1
  diskFailure:
    path: /mnt/data/disk-read
...
```

</details>

<p align="center">
hello-dark
</p>

> **After**

<details>

``` yaml
---
apiVersion: chaos.datadoghq.com/v1beta1
kind: Disruption
metadata:
  name: disk-failure
  namespace: chaos-demo
  annotations:
    chaos.datadoghq.com/environment: "lima"
spec:
  level: pod
  selector:
    app: demo-curl
  count: 1
  diskFailure:
    paths: 
      - /mnt/data/disk-read
      - /mnt/data/disk-write
 ...
```

</details>

<p align="center">

</p>

## Code Quality Checklist

- [x] The documentation is up to date.
- [x] My code is sufficiently commented and passes continuous integration checks.
- [x] I have signed my commit (see [Contributing Docs](../CONTRIBUTING.md)).

## Testing

- [ ] I leveraged continuous integration testing
    - [ ] by depending on existing `unit` tests or `end-to-end` tests.
    - [ ] by adding new `unit` tests or `end-to-end` tests.
- [x] I manually tested the following steps:
    - [x] locally.
    - [ ] as a canary deployment to a cluster.

### CODE REVIEW

File: api/v1beta1/disk_failure.go [hunk 1]
```diff
@@ -18,6 +18,50 @@ type OpenatSyscallSpec struct {
...
+func (s *DiskFailureSpec) Validate() error {
+	for _, path := range s.Paths {
+		path := strings.TrimSpace(path)
+
+		if path == "" {
+			return fmt.Errorf("the path of the disk failure disruption must not be empty")
+		}
+
+		if len(path) > MaxDiskPathCharacters {
+			return fmt.Errorf("the path of the disk failure disruption must not exceed %d characters", MaxDiskPathCharacters)
```
Comment (luphaz) [inline api/v1beta1/disk_failure.go:42]:
```suggestion
			return fmt.Errorf("the path of the disk failure disruption must not exceed %d characters, found %d", MaxDiskPathCharacters, len(path))
```

File: api/v1beta1/disk_failure.go [hunk 2]
```diff
@@ -18,6 +18,50 @@ type OpenatSyscallSpec struct {
 	ExitCode string `json:"exitCode"`
 }
 
+// DiskFailureSpec represents a disk failure disruption
+type DiskFailureSpec struct {
+	// +kubebuilder:validation:Required
+	// +ddmark:validation:Required=true
+	Paths []string `json:"paths"`
```
Comment (luphaz) [inline api/v1beta1/disk_failure.go:25]: Are we ok to just drop a field?
It seems it's a breaking change for me 🤔 

Maybe it's ok if we did not released this as part of a numbered release already 🤷

File: injector/disk_failure.go [hunk 3]
```diff
@@ -101,7 +101,13 @@ func (i *DiskFailureInjector) Inject() error {
 		exitCode = i.spec.OpenatSyscall.GetExitCodeInt()
 	}
 
-	return i.config.Cmd.Run(pid, i.spec.Path, exitCode)
+	for _, path := range i.spec.Paths {
+		if err := i.config.Cmd.Run(pid, path, exitCode); err != nil {
```
Comment (luphaz) [inline injector/disk_failure.go:105]: You may want to have a look at the BackgroundCmd struct [here](https://github.com/DataDog/chaos-controller/blob/323596d971a465428ac8aa5180e7a77f900af0aa/command/command.go#L112)

Used [here](https://github.com/DataDog/chaos-controller/blob/323596d971a465428ac8aa5180e7a77f900af0aa/injector/dns_disruption.go#L130)

It's not directly part to this review though.

A comment may be necessary as I had to check the implementation to confirm it was not a synchronous call.

**Review by luphaz — APPROVED** [review]

File: injector/disk_failure.go (hunk 3 above)
Comment (aymericDD) [reply injector/disk_failure.go:105]: Interesting 👍 I will take a look!

File: command/command.go [hunk 4]
```diff
@@ -197,7 +198,7 @@ func (w *backgroundCmd) KeepAlive() {
 
 			<-w.safeTicker().C
 
-			process, err := w.processManager.Find(w.pid)
+			proc, err := w.processManager.Find(w.pid)
```
Comment (aymericDD) [inline command/command.go:201]: Remove shadowing.

File: injector/disk_failure.go [hunk 5]
```diff
@@ -102,8 +63,22 @@ func (i *DiskFailureInjector) Inject() error {
...
+			args = append(args, "-f", path)
+		}
+
+		if exitCode != 0 {
+			args = append(args, "-c", fmt.Sprintf("%v", exitCode))
+		}
+
+		cmd := i.config.CmdFactory.NewCmd(context.Background(), EBPFDiskFailureCmd, args)
+
+		bgCmd := command.NewBackgroundCmd(cmd, i.config.Log, i.config.ProcessManager)
```
Comment (aymericDD) [inline injector/disk_failure.go:78]: I don't find a way to mock this part 😢

Comment (luphaz) [reply injector/disk_failure.go:78]: What I wanted to test for the CPU was:
* it does not matter how do you build your command, in the end I want to ensure "start" was called

https://github.com/DataDog/chaos-controller/blob/323596d971a465428ac8aa5180e7a77f900af0aa/injector/factory.go#L77

Hence I created a factory that will return a background command, factory that can be mocked for testing:
* https://github.com/DataDog/chaos-controller/blob/323596d971a465428ac8aa5180e7a77f900af0aa/injector/cpu_pressure_test.go#L43C23-L43C23

Background command hence can be a mock

If you really want to "mock" the creation of a command, you should provide a func as a argument that default to the `NewBackgroundCmd` as an example.

It really depends on how much you want to test and how.

Sometimes, doing only "component" testing is enough.

It's up to you to define at which granularity you need to go.

**Review by luphaz — APPROVED** [review]

**Review by taihuynh167 — APPROVED** [review]

**Pull request merged**